### PR TITLE
topsql: a centralized place to generate tipb report data

### PR DESCRIPTION
**Pull request by zhongzc**
Signed-off-by: Zhenchi <zhongzc_arch@outlook.com>

### What problem does this PR solve?

Part of #29020

Problem Summary:

Many datasinks will be interested in `ReportData`. It's wasted to convert them from the original shape to the tipb shape multiple times.

### What is changed and how it works?

Change fields in `ReportData` to be in tipb shape. Converting them before sending them to datasinks.

### Check List

Tests 

- [x] Unit test
- [x] Integration test

### Release note

```release-note
None
```

### CODE REVIEW

File: util/topsql/reporter/reporter.go [hunk 1]
```diff
@@ -523,11 +519,53 @@ func (tsr *RemoteTopSQLReporter) getReportData(collected collectedData) ReportDa
...
+			RecordListCpuTimeMs:    record.CPUTimeMsList,
+			SqlDigest:              record.SQLDigest,
+			PlanDigest:             record.PlanDigest,
+		})
+	}
+
+	sqlMap.Range(func(key, value interface{}) bool {
+		meta := value.(SQLMeta)
+		res.SQLMetas = append(res.SQLMetas, &tipb.SQLMeta{
+			SqlDigest:     []byte(key.(string)),
```
Comment (breeswish) [inline util/topsql/reporter/reporter.go:548]: is it by design that they are stored as binary strings?

File: util/topsql/reporter/reporter.go [hunk 2]
```diff
@@ -451,29 +456,14 @@ type collectedData struct {
...
 // ReportData contains data that reporter sends to the agent
 type ReportData struct {
-	// collectedData contains the topN collected records and the `others` record which aggregation all records that is out of Top N.
-	collectedData     []*dataPoints
-	normalizedSQLMap  *sync.Map
-	normalizedPlanMap *sync.Map
+	// CPUTimeRecords contains the topN collected records and the `others` record which aggregation all records that is out of Top N.
+	CPUTimeRecords []*tipb.CPUTimeRecord
+	SQLMetas       []*tipb.SQLMeta
+	PlanMetas      []*tipb.PlanMeta
```
Comment (breeswish) [inline util/topsql/reporter/reporter.go]: It would be better to be `[]Foo` instead of `[]*Foo`.

File: util/topsql/reporter/single_target.go [hunk 3]
```diff
@@ -180,111 +175,105 @@ func (ds *SingleTargetDataSink) doSend(ctx context.Context, addr string, data Re
...
-		cnt++
-		planMeta := &tipb.PlanMeta{
-			PlanDigest:     []byte(key.(string)),
-			NormalizedPlan: planDecoded,
-		}
-		if err = stream.Send(planMeta); err != nil {
-			return false
+
+	for _, meta := range planMetas {
+		if err = stream.Send(meta); err != nil {
```
Comment (breeswish) [inline util/topsql/reporter/single_target.go]: Do we need some kind of batch mechanism or flow control? As now there can be a lot. @crazycs520

**Review by breeswish — COMMENTED** [review]
the rest LGTM

**Review by breeswish — APPROVED** [review]

File: util/topsql/reporter/reporter.go (hunk 1 above)
Comment (zhongzc) [reply util/topsql/reporter/reporter.go:548]: Don't know. It has been like that for a long while.

**Review by crazycs520 — APPROVED** [review]

File: util/topsql/reporter/single_target.go (hunk 3 above)
Comment (breeswish) [reply util/topsql/reporter/single_target.go]: https://github.com/pingcap/tidb/issues/30787

Opened an issue to track that. Without any tracking seems that the work can be easily forgotten.

**Pull request merged**